presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java

```diff
@@ -101,12 +103,32 @@ private SqlInvokedFunction createSqlInvokedFunction(CreateFunction statement, Me
 
         if (statement.getBody() instanceof Return) {
             Expression bodyExpression = ((Return) statement.getBody()).getExpression();
-            Type bodyType = analysis.getType(bodyExpression);
-
-            if (!bodyType.equals(metadata.getType(returnType))) {
-                // Casting is safe-here, since we have verified that the actual type of the body is coercible to declared return type.
-                body = new Return(new Cast(bodyExpression, statement.getReturnType()));
+            if (!analysis.getType(bodyExpression).equals(metadata.getType(returnType))) {
+                // Casting is safe here, since we have verified at analysis time that the actual type of the body is coercible to declared return type.
+                bodyExpression = new Cast(bodyExpression, statement.getReturnType());
             }
+
+            // Coerce expressions in body if necessary
+            bodyExpression = ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Void>()
```

Comment:
Nits: I found it more logical to have this before the return cast rather than after.

Reply:
Ok, done.